File: D-Circumferences.py
from typing import List, Tuple


class Circumferences:
    def __init__(self, n: int, sx: int, sy: int, tx: int, ty: int, circles: List[Tuple[int, int, int]]) -> None:
        self.n = n
        self.s = (sx, sy)
        self.t = (tx, ty)
        self.circles = circles

    def create_graphs(self) -> List[List[int]]:
        graphs = [[] for _ in range(self.n)]

        for i in range(self.n):
            for j in range(i + 1, self.n):
                circle_i = self.circles[i]
                circle_j = self.circles[j]
                if (circle_i[2] + circle_j[2]) ** 2 < (circle_i[0] - circle_j[0]) ** 2 + (
                        circle_i[1] - circle_j[1]) ** 2:
                    continue
                if (circle_i[2] - circle_j[2]) ** 2 > (circle_i[0] - circle_j[0]) ** 2 + (
                        circle_i[1] - circle_j[1]) ** 2:
                    continue
                graphs[i].append(j)

        return graphs

    def is_circle_on_the_point(self, x: int, y: int, circle_num: int) -> bool:
        circle = self.circles[circle_num]
        if circle[2] ** 2 == (x - circle[0]) ** 2 + (y - circle[1]) ** 2:
            return True
        return False

    def find_circle_on_the_point(self, point: Tuple[int, int]) -> int:
        for circle_num in range(self.n):
            if self.is_circle_on_the_point(point[0], point[1], circle_num):
                return circle_num


class UnionFind:
    def __init__(self, n: int) -> None:
        self.n = n
        self.parents = [-1] * n

    def find(self, x: int) -> int:
        if self.parents[x] < 0:
            return x
        self.parents[x] = self.find(self.parents[x])
        return self.parents[x]

    def union(self, x: int, y: int) -> None:
        x = self.find(x)
        y = self.find(y)

        if x == y:
            return

        if self.parents[x] > self.parents[y]:
            x, y = y, x

        self.parents[x] += self.parents[y]
        self.parents[y] = x

    def size(self, x: int) -> int:
        return -self.parents[self.find(x)]

    def same(self, x: int, y: int) -> bool:
        return self.find(x) == self.find(y)


def main_judge(n: int, sx: int, sy: int, tx: int, ty: int, circles: List[Tuple[int, int, int]]) -> bool:
    circumferences = Circumferences(n, sx, sy, tx, ty, circles)
    graphs = circumferences.create_graphs()
    union = UnionFind(n)
    for i, circle_list in enumerate(graphs):
        for circle_num in circle_list:
            union.union(i, circle_num)

    if union.same(circumferences.find_circle_on_the_point((sx, sy)), circumferences.find_circle_on_the_point((tx, ty))):
        return True
    return False


def main():
    n = int(input())
    sx, sy, tx, ty = map(int, input().split())
    circles = []
    for _ in range(n):
        x, y, r = map(int, input().split())
        circles.append((x, y, r))
    if main_judge(n, sx, sy, tx, ty, circles):
        print('Yes')
    else:
        print('No')


if __name__ == '__main__':
    main()

File: D-Circumferences.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

var sc = bufio.NewScanner(os.Stdin)
var rdr = bufio.NewReaderSize(os.Stdin, 1000000)

func stringInput() string {
	sc.Scan()
	return sc.Text()
}

func intInput() int {
	sc.Scan()
	x, err := strconv.Atoi(sc.Text())
	if err != nil {
		panic(err)
	}
	return x
}

func intSliceInput(n int) []int {
	sl := []int{}
	for i := 0; i < n; i++ {
		x := intInput()
		sl = append(sl, x)
	}
	return sl
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func initSlSl(y, x int) [][]int {
	sl := make([][]int, y)
	if x == -1 {
		for i := 0; i < y; i++ {
			sl[i] = []int{}
		}
		return sl
	}
	for i := 0; i < y; i++ {
		sl[i] = make([]int, x)
	}
	return sl
}

func init() {
	sc.Split(bufio.ScanWords)
}

type UnionFind struct {
	parSlice  []int
	sizeSlice []int
}

func createStruct(n int) *UnionFind {
	uni := new(UnionFind)
	for i := 0; i < n; i++ {
		uni.parSlice = append(uni.parSlice, -1)
		uni.sizeSlice = append(uni.sizeSlice, 1)
	}
	return uni
}

func (uni *UnionFind) root(x int) int {
	if uni.parSlice[x] == -1 {
		return x
	} else {
		uni.parSlice[x] = uni.root(uni.parSlice[x])
		return uni.parSlice[x]
	}
}

func (uni *UnionFind) isSame(x, y int) bool {
	return uni.root(x) == uni.root(y)
}

func (uni *UnionFind) unite(x, y int) bool {
	x = uni.root(x)
	y = uni.root(y)
	if x == y {
		return false
	}
	if uni.sizeSlice[x] < uni.sizeSlice[y] {
		x, y = y, x
	}
	uni.parSlice[y] = x
	uni.sizeSlice[x] += uni.sizeSlice[y]
	return true
}

func (uni *UnionFind) size(x int) int {
	return uni.sizeSlice[uni.root(x)]
}

func pow(x, y int) int {
	return int(math.Pow(float64(x), float64(y)))
}

func createGraphs(circles [][]int) [][]int {
	n := len(circles)
	graphs := make([][]int, n)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			circleI := circles[i]
			circleJ := circles[j]
			if pow(circleI[2]+circleJ[2], 2) < pow(circleI[0]-circleJ[0], 2)+pow(circleI[1]-circleJ[1], 2) {
				continue
			}
			if pow(circleI[2]-circleJ[2], 2) > pow(circleI[0]-circleJ[0], 2)+pow(circleI[1]-circleJ[1], 2) {
				continue
			}
			graphs[i] = append(graphs[i], j)
		}
	}
	return graphs
}

func isCircleOnThePoint(circle []int, x int, y int) bool {
	if pow(circle[2], 2) == pow(x-circle[0], 2)+pow(y-circle[1], 2) {
		return true
	}
	return false
}

func findCircleOnThePoint(point [2]int, circles [][]int) int {
	result := -1
	for i, circle := range circles {
		if isCircleOnThePoint(circle, point[0], point[1]) {
			result = i
			break
		}
	}
	return result
}

func mainJudge(circles [][]int, n, sx, sy, tx, ty int) bool {
	graphs := createGraphs(circles)
	uni := createStruct(n)
	for i, nextCircle := range graphs {
		for _, circle := range nextCircle {
			uni.unite(i, circle)
		}
	}
	startCircle := findCircleOnThePoint([2]int{sx, sy}, circles)
	endCircle := findCircleOnThePoint([2]int{tx, ty}, circles)
	return uni.isSame(startCircle, endCircle)
}

func main() {
	n := intInput()
	sx, sy, tx, ty := intInput(), intInput(), intInput(), intInput()
	circles := make([][]int, n)
	for i := 0; i < n; i++ {
		xyr := intSliceInput(3)
		circles[i] = xyr
	}

	if mainJudge(circles, n, sx, sy, tx, ty) {
		fmt.Println("Yes")
	} else {
		fmt.Println("No")
	}

}
